### PR TITLE
Add a constructor for ActivationCache from a list of Caches

### PR DESCRIPTION
**Pull request by afspies**
I wanted to combine many activation caches into one big one, and created a constructor to that end.

Not sure if this is elegant / something you want - feel free to reject. If you do think this is useful, then perhaps it is worth adjusting the `model.run_with_cache()` function to accept a cache which it writes into (i.e. just calling `return ActivationCache.from_caches([cache_computed, cache_passed_in], in_place=True)`)

### CODE REVIEW

File: transformer_lens/ActivationCache.py [hunk 1]
```diff
@@ -43,6 +43,39 @@ def __init__(
         self.has_embed = "hook_embed" in self.cache_dict
         self.has_pos_embed = "hook_pos_embed" in self.cache_dict
 
+    @classmethod
+    def from_caches(cls, cache_list: List[ActivationCache], in_place: Bool = True) -> ActivationCache:
+        # Constructor that combines a list of caches into a big one
```
Comment (neelnanda-io) [inline transformer_lens/ActivationCache.py:48]: Can you add a docstring? In particular, what `in_place` does should be documented

File: transformer_lens/ActivationCache.py [hunk 2]
```diff
@@ -43,6 +43,39 @@ def __init__(
...
+        assert all(
+            cache.has_batch_dim for cache in cache_list
+        ), "Cannot stack caches without batch dimensions"
+
+        # Concat a dict of lists once rather than concatting many times
+        cache_dict = {k:[v] for k,v in cache_list[0].cache_dict.items()}
+
+        if not in_place:
+            cache_dict = deepcopy(cache_dict)
+        model = cache_list[0].model.name
```
Comment (neelnanda-io) [inline transformer_lens/ActivationCache.py:60]: Rename to `model_name`

File: transformer_lens/ActivationCache.py [hunk 3]
```diff
@@ -43,6 +43,39 @@ def __init__(
...
+            
+            for key in cache_dict:
+                # cache_dict[key] = torch.cat((cache_dict[key], activation_cache.cache_dict[key]), dim=0)
+                cache_dict[key].append(activation_cache.cache_dict[key])
+
+        cache_dict = {k:torch.cat(v, dim=0) for k,v in cache_dict.items()}
+        if in_place:
+            [cache.cache_dict.clear() for cache in cache_list] 
+
+        return cls(cache_dict, model, has_batch_dim=True)
```
Comment (neelnanda-io) [inline transformer_lens/ActivationCache.py:77]: You want the actual model, not the model's name

File: transformer_lens/ActivationCache.py (hunk 2 above)
Comment (neelnanda-io) [reply transformer_lens/ActivationCache.py:60]: Also this doesn't work. You maybe want `model.cfg.model_name`?

Comment (neelnanda-io) [reply transformer_lens/ActivationCache.py:60]: Not a big deal either way

File: transformer_lens/ActivationCache.py [hunk 4]
```diff
@@ -43,6 +43,39 @@ def __init__(
...
+        
+        assert len(cache_list) > 0, "Cannot stack an empty list of caches"
+        assert all(
+            cache.has_batch_dim for cache in cache_list
+        ), "Cannot stack caches without batch dimensions"
+
+        # Concat a dict of lists once rather than concatting many times
+        cache_dict = {k:[v] for k,v in cache_list[0].cache_dict.items()}
+
+        if not in_place:
```
Comment (neelnanda-io) [inline transformer_lens/ActivationCache.py:58]: Unnecessary, you're concatenating them anyway

File: transformer_lens/ActivationCache.py [hunk 5]
```diff
@@ -43,6 +43,39 @@ def __init__(
...
+            assert has_embed == activation_cache.has_embed, "Cannot stack caches with different embed hooks"
+            assert has_pos_embed == activation_cache.has_pos_embed, "Cannot stack caches with different pos_embed hooks"
+            
+            for key in cache_dict:
+                # cache_dict[key] = torch.cat((cache_dict[key], activation_cache.cache_dict[key]), dim=0)
+                cache_dict[key].append(activation_cache.cache_dict[key])
+
+        cache_dict = {k:torch.cat(v, dim=0) for k,v in cache_dict.items()}
+        if in_place:
+            [cache.cache_dict.clear() for cache in cache_list]
```
Comment (neelnanda-io) [inline transformer_lens/ActivationCache.py:75]: Is memory management actually an issue here? If so, you'll probably get the best results by going through each activation type, concatenating all of those across all caches, and then deleting the component activations.

Also, moving things to the CPU! Helps a lot

Comment (neelnanda-io) [reply transformer_lens/ActivationCache.py:75]: IMO you can just not care though.

**Review by neelnanda-io — CHANGES_REQUESTED** [review]
Thanks! Seems like a reasonable feature, my most important requested changes are a docstring and fixing the `model.name` part.

**Comment by neelnanda-io** [comment]
I don't think this is *that *useful, so I don't want to merge it into
`run_with_cache` (I already have an ugh at how bloated that function is,
and how it's mutated beyond HookedRootModule, but is technically a subclass
of it)

File: transformer_lens/ActivationCache.py (hunk 2 above)
Comment (afspies) [reply transformer_lens/ActivationCache.py:60]: Thanks for noticing that!

**Review by afspies — COMMENTED** [review]
Have made the requested changes

**Comment by jbloomAus** [comment]
Closed based on the above conversation (ActivationCache being very bloated already).

**Pull request closed without merging**